### ISSUE TITLE
How to concatenate 4 images into a grid

### ISSUE THREAD

**Issue by DPG7332** [chaiNNer-org/chaiNNer#1441]
Objective:
Create a Chainner script that will automatically combine 4 images into a 4x4 grid

I previously split an image up into 4 equal pieces, (1.jpg, 2.jpg, 3.jpg and 4.jpg) where 1.jpg is the top left corner, 2.jpg is the top right, 3.jpg is the bottom left, and 4.jpg is the bottom right. I want to write a chainner script that will combine all four pieces back into one image. This is rather simple when I load each image separately using the Load Image node, but I don't want to do that. Instead I want Chainner to iterate through folders and do it automatically.  Doesn't seem possible since I can't store the loaded images in an array. Can anyone help?

Here is the project that does it manually

<img width="996" alt="image" src="https://user-images.githubusercontent.com/112595751/209767718-d30fc589-79c2-4df5-9f35-cc3099c51d3f.png">

And here is a Python script I wrote that works.  However I want to do this in Chainner. 

`# Import the necessary libraries
import os
import numpy as np
from PIL import Image

# Define the path of the parent folder
parent_folder_path = "CombineImages"

# Loop through all subfolders in the parent folder
for folder_name in os.listdir(parent_folder_path):
    folder_path = os.path.join(parent_folder_path, folder_name)
    if os.path.isdir(folder_path):
        # Define the paths to the four image pieces
        piece1_path = os.path.join(folder_path, "1.jpg")
        piece2_path = os.path.join(folder_path, "2.jpg")
        piece3_path = os.path.join(folder_path, "3.jpg")
        piece4_path = os.path.join(folder_path, "4.jpg")

        # Load the image pieces as NumPy arrays
        piece1 = np.array(Image.open(piece1_path))
        piece2 = np.array(Image.open(piece2_path))
        piece3 = np.array(Image.open(piece3_path))
        piece4 = np.array(Image.open(piece4_path))

        # Concatenate the image pieces along the vertical axis to form the top half of the reassembled image
        top_half = np.concatenate((piece1, piece2), axis=1)

        # Concatenate the image pieces along the vertical axis to form the bottom half of the reassembled image
        bottom_half = np.concatenate((piece3, piece4), axis=1)

        # Concatenate the top and bottom halves along the horizontal axis to form the reassembled image
        reassembled_image = np.concatenate((top_half, bottom_half), axis=0)

        # Save the reassembled image with the word "stitched" added to the file name
        output_path = piece1_path.split(".")[0] + "_stitched.jpg"
        Image.fromarray(reassembled_image).save(output_path)`

**Comment by 0x4E69676874466F78** [comment]
For very large images you can use libvips which has bindings to python https://pypi.org/project/pyvips/

**Comment by joeyballentine** [comment]
This will be trivial once we rewrite iterators. For now, you'd either have to do this manually like you said, or you can actually use the sprite sheet iterators to do both the splitting and combining steps (even though this isn't a sprite sheet). 

If you don't mind me asking, what's the reason you have these separated in the first place?

**Comment by DPG7332** [comment]
Cool. looking forward to the rewrite, when is that planned for?

I split them up initially to solve a problem. The image I was trying to 4x upscale is 8000x8000px and apparently to big because it was causing Chainner to crash. To fix the issue I modified the Chainner project to first split the image into 4 equal pieces, upscale each piece, stitch all 4 upscaled images back together  before saving the big  32,000 x 32,000 px file. This worked well, but the entire process was pretty hefty, so i decided to split up the tasks. The first step is to split up the images into pieces, upscale them, then save each upscaled piece. The second step that I am building now in my workflow is to stitch them back up and save the big file. I'm paying for a VM with GPU  to do the upscaling so want to make this process as efficient as possible. If I can execute the stitch up tasks in Python and perhaps do it on my macbook M1 air instead of the VM, that would be more cost efficient. Perhaps there is another better way?

In any case, really love Chainner, keep up the amazing work.

**Comment by joeyballentine** [comment]
In what way was it causing chainner to crash? We actually perform tiling (splitting and merging) already for upscaling, though the caching is done in memory. Are you running out of RAM?

**Comment by DPG7332** [comment]
I suspect it was running out of RAM, although the error message was not clear on that. I can try it again to collect that information but that seemed to be the problem.

**Comment by joeyballentine** [comment]
That would be helpful yes, if you don't mind

**Comment by DPG7332** [comment]
Here is the error I got, happened about 15 min into the job.


<img width="1433" alt="image" src="https://user-images.githubusercontent.com/112595751/209973646-5e3ed8f0-2086-4cb9-9243-0ee04c2b38d6.png">

**Comment by RunDevelopment** [comment]
Yes, you ran out of RAM. 12GB for just the output image is no joke, and this doesn't include the RAM needed to do the actual upscaling. So unless you have around 32GB of RAM, this won't work.

That being said, 32k images are kinda at the limit of what chainner can do anyway. I would recommend excluding that images this large from the iteration by moving them into a different directory.

**Comment by DPG7332** [comment]
Ok got it. Since the tiling approach I used does work with a 32k output file, I'll stick with that for now.
